frost/src/utils.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{preprocess::SigningCommitment, GroupPublicKey};
+
+use console::algorithms::{Hash, HashToScalar, Poseidon4};
+use snarkvm_curves::{AffineCurve, ProjectiveCurve};
+use snarkvm_fields::{Field, One, PrimeField, ToConstraintField, Zero};
+
+use anyhow::{anyhow, Result};
+use std::{
+    collections::HashMap,
+    ops::{Mul, Sub},
+};
+
+/// Calculate the Lagrange coefficient for a given participant index.
+pub fn calculate_lagrange_coefficients<G: AffineCurve>(
+    participant_index: u64,
+    all_participant_indices: &[u64],
```

Comment:
looking at the callsites, this could be an `impl Iterator<Item = u64>` to avoid `Vec` allocations